rllib/core/rl_trainer/rl_trainer.py

```diff
@@ -0,0 +1,142 @@
+        self.optimizer_class = optimizer_class
+        self.optimizer_config = optimizer_config
+        self.distributed = distributed
+        self.debug = debug
+
+    @staticmethod
+    def on_after_compute_gradients(
+        gradients_dict: Mapping[str, Any]
+    ) -> Mapping[str, Any]:
+        """Called after gradients have been computed.
+
+        Args:
+            gradients_dict (Mapping[str, Any]): A dictionary of gradients.
+
+        Note the relative order of operations looks like this:
+            fwd_out = forward_train(batch)
+            loss = compute_loss(batch, fwd_out)
+            gradients = compute_gradients(loss)
+            ---> post_processed_gradients = on_after_compute_gradients(gradients)
+            apply_gradients(post_processed_gradients)
+
+        Returns:
+            Mapping[str, Any]: A dictionary of gradients.
+        """
+        return gradients_dict
+
+    def compile_results(
+        self,
+        batch: NestedDict,
+        fwd_out: Mapping[str, Any],
+        postprocessed_loss: Mapping[str, Any],
+        post_processed_gradients: Mapping[str, Any],
+    ) -> Mapping[str, Any]:
+        """Compile results from the update.
+
+        Args:
+            batch: The batch that was used for the update.
+            fwd_out: The output of the forward train pass.
+            postprocessed_loss: The loss after postprocessing.
+            post_processed_gradients: The gradients after postprocessing.
+
+        Returns:
+            A dictionary of results.
+        """
+        loss_numpy = convert_to_numpy(postprocessed_loss)
+        batch = convert_to_numpy(batch)
+        return {
+            "avg_reward": batch["rewards"].mean(),
+            **loss_numpy,
+        }
+
+    @abc.abstractmethod
+    def update(self, batch: SampleBatch) -> Mapping[str, Any]:
+        """Perform an update on this Trainer.
+
+        Args:
+            batch: A batch of data.
+
+        Returns:
+            A dictionary of results.
+        """
+
+    @abc.abstractmethod
+    def compute_gradients(
+        self, loss: Union[TensorType, Mapping[str, Any]]
+    ) -> Mapping[str, Any]:
+        """Perform an update on self._module.
+
+        For example compute and apply gradients to self._module if
+        necessary.
+
+        Args:
+            loss: variable(s) used for optimizing self._module.
+
+        Returns:
+            A dictionary of extra information and statistics.
+        """
+
+    def set_state(self, state: Mapping[str, Any]) -> None:
+        """Set the state of the trainer."""
+        self._rl_optimizer.set_state(state.get("optimizer_state", {}))
+        self._module.set_state(state.get("module_state", {}))
+
+    def get_state(self) -> Mapping[str, Any]:
+        """Get the state of the trainer."""
+        return {
+            "module_state": self._module.get_state(),
+            "optimizer_state": self._rl_optimizer.get_state(),
+        }
+
+    def init_trainer(self) -> None:
```

Comment:
Should `init_trainer` be a public function on RLTrainer? should it not be called inside the constructor of the class at the time of construction?

Reply:
no this is a limitation of ray train. The module initialization needs to happen after init.

Reply:
Oh I see. So the actor needs to get initialized first to get the backend stuff. makes sense. Is there any established pattern for these types of actors? I was thinking maybe we should have a `.start()` , `.stop()` api?